### PR TITLE
Item migration [RDY]

### PR DESCRIPTION
**Pull request by mugling**
Factored out of #16568 and provides mechanism for migrating legacy saves.

Includes ability to rename item `id`, set `charges` and/or `flags`. Suggestions as to other useful functionality welcomed.

### CODE REVIEW

**Comment by Mshock777** [comment]
A bit offtopic: maybe we should mainline Xotto tileset, if it isn't, it won't get any maintenance like this.

**Comment by illi-kun** [comment]
> we should mainline Xotto tileset

It definitely worth the efforts.

**Comment by mugling** [comment]
Three demonstration items are `m14a`, `saiga_sawn` and `bag_bundle_10`.

To test create a save with these items then compile this branch and load the save.

* M1A should be unchanged (apart from the `id`)
* Saiga should be converted to the base type plus sawn barrel mod
* Bundle of plastic bags should be destroyed

File: src/item.cpp [hunk 1]
```diff
@@ -90,9 +90,7 @@ std::string const& rad_badge_color(int const rad)
 
 light_emission nolight = {0, 0, 0};
 
-// Returns the default item type, used for the null item (default constructed),
-// the returned pointer is always valid, it's never cleared by the @ref Item_factory.
-static const itype *nullitem()
+const itype * item::nullitem()
```
Comment (kevingranade) [inline src/item.cpp]: Nit. Style regression, pointer dereference operator is right-aligned.

File: src/item_factory.cpp [hunk 2]
```diff
@@ -1495,6 +1506,53 @@ void Item_factory::load_item_category(JsonObject &jo)
...
+    }
+    return id;
+}
+
+void Item_factory::migrate_item( const itype_id& id, item& obj )
+{
+    auto iter = migrations.find( id );
+    if( iter != migrations.end() ) {
+        if( iter->second.destroy ) {
+            obj.type = item::nullitem();
```
Comment (BevapDin) [inline src/item_factory.cpp]: Outright destroying the item makes the whole loading code more complicated (you had to introduce several `is_null` checks) and player will report this as a bug ("item just vanished upon loading my save!").

Why not just convert it to a standard note item ("obsolete item"), maybe with a proper description explaining that this item had been removed from the game. This avoids the null check (still a valid item), and allows the player to discover that change.

File: src/json.cpp [hunk 3]
```diff
@@ -12,6 +12,8 @@
 #include <vector>
 #include <bitset>
 
+#include "item.h"
```
Comment (BevapDin) [inline src/json.cpp]: This disables the chkjson binary. It only contains the chkjson.cpp and json.cpp, but does not include the other source files (and therefor the reference to `item` will be missing).

**Comment by mugling** [comment]
I've backed that item deletion part for now as that's whats causing the current holdup and it's not required so as to be able to rebase #16568. We can revisit item deletion at a later date.

**Comment by mugling** [comment]
Of my open PR's mark this is as the next most relevant

**Pull request merged**